src/processors/file-handler.js

```diff
@@ -167,12 +83,119 @@ export function isKeplerGlMap(json) {
   );
 }
 
-export function determineJsonProcess({dataset, format}, defaultProcessor) {
-  if (isKeplerGlMap(dataset)) {
-    return processKeplerglJSON;
+export async function* makeProgressIterator(asyncIterator, info) {
+  const startTime = Date.now();
+  let rowCount = 0;
+
+  for await (const batch of asyncIterator) {
+    const rowCountInBatch = (batch.data && batch.data.length) || 0;
+    rowCount += rowCountInBatch;
+    const elapsedMs = Date.now() - startTime;
+    const percent = batch.bytesUsed / info.size;
+
+    // Update progress object
+    const progress = {
+      elapsedMs,
+      rowCount,
+      rowCountInBatch,
+      percent
+    };
+
+    yield {...batch, progress};
+  }
+}
+
+// eslint-disable-next-line complexity
+async function* readBatch(asyncIterator, fileName) {
+  let result = null;
+  let batches = [];
+
+  for await (const batch of asyncIterator) {
+    // Last batch will have this special type and will provide all the root
+    // properties of the parsed document.
+    if (batch.batchType === BATCH_TYPE.FINAL_RESULT) {
+      if (batch.container) {
+        result = {...batch.container};
+      }
+      // Set the streamed data correctly is Batch json path is set
+      // and the path streamed is not the top level object (jsonpath = '$')
+      if (batch.jsonpath && batch.jsonpath.length > 1) {
+        const streamingPath = new _JSONPath(batch.jsonpath);
+        streamingPath.setFieldAtPath(result, batches);
+      } else {
+        // The streamed object is a ROW JSON-batch
+        result = batches;
+      }
+    } else {
+      for (let i = 0; i < batch.data.length; i++) {
+        batches.push(batch.data[i]);
+      }
+    }
+
+    yield {
+      ...batch,
+      header: batch.schema ? Object.keys(batch.schema) : null,
+      fileName,
+      // if dataset is CSV, data is set to the raw batches
+      data: result ? result : batches
+    };
   }
+}
 
-  return defaultProcessor;
+export async function readFileInBatches({file, fileCache = []}) {
+  const batchIterator = await parseInBatches(
+    file,
+    [JSONLoader, CSVLoader],
+    {
+      csv: CSV_LOADER_OPTIONS,
+      json: JSON_LOADER_OPTIONS,
+      metadata: true
+    },
+    file.name
+  );
+  const progressIterator = makeProgressIterator(batchIterator, {size: file.size});
+  return readBatch(progressIterator, file.name);
+}
+
+export function processFileData({content, fileCache}) {
+  return new Promise((resolve, reject) => {
+    const {data, header} = content;
+
+    let format;
+    let processor;
+    if (isCsvRows(data)) {
```

Comment:
If we need to know if this was originally CSV then we can use the metadata batch instead of sniffing like this. The idea with loaders.gl is that we can add additional streaming parsers do we need to keep track of what type things were originally?